numba/stencilparfor.py

```diff
@@ -0,0 +1,622 @@
+            arg_typs = (slice_type, types.intp,)
+        _globals = self.func_ir.func_id.func.__globals__
+        f_ir = compile_to_numba_ir(f, _globals, self.typingctx, arg_typs,
+                                    self.typemap, self.calltypes)
+        _, block = f_ir.blocks.popitem()
+        replace_arg_nodes(block, args)
+        new_index = block.body[-2].value.value
+        out_nodes.extend(block.body[:-2])  # ignore return nodes
+        return new_index
+
+def get_stencil_blocks(sf, typingctx, args, scope, loc, input_dict, typemap,
+                                                                    calltypes):
+    """get typed IR from stencil bytecode
+    """
+    from numba.targets.cpu import CPUContext
+    from numba.targets.registry import cpu_target
+    from numba.annotations import type_annotations
+    from numba.compiler import type_inference_stage
+
+    # get untyped IR
+    stencil_func_ir = sf.kernel_ir.copy()
+    # copy the IR nodes to avoid changing IR in the StencilFunc object
+    stencil_blocks = copy.deepcopy(stencil_func_ir.blocks)
+    stencil_func_ir.blocks = stencil_blocks
+
+    # get typed IR with a dummy pipeline (similar to test_parfors.py)
+    targetctx = CPUContext(typingctx)
+    with cpu_target.nested_context(typingctx, targetctx):
+        tp = DummyPipeline(typingctx, targetctx, args, stencil_func_ir)
+
+        numba.rewrites.rewrite_registry.apply(
+            'before-inference', tp, tp.func_ir)
+
+        tp.typemap, tp.return_type, tp.calltypes = type_inference_stage(
+            tp.typingctx, tp.func_ir, tp.args, None)
+
+        type_annotation = type_annotations.TypeAnnotation(
+            func_ir=tp.func_ir,
+            typemap=tp.typemap,
+            calltypes=tp.calltypes,
+            lifted=(),
+            lifted_from=None,
+            args=tp.args,
+            return_type=tp.return_type,
+            html_output=numba.config.HTML)
+
+    # make block labels unique
+    stencil_blocks = ir_utils.add_offset_to_labels(stencil_blocks,
+                                                        ir_utils.next_label())
+    min_label = min(stencil_blocks.keys())
+    max_label = max(stencil_blocks.keys())
+    ir_utils._max_label = max_label
+
+    if config.DEBUG_ARRAY_OPT == 1:
+        print("Initial stencil_blocks")
+        ir_utils.dump_blocks(stencil_blocks)
+
+    # rename variables,
+    var_dict = {}
+    for v, typ in tp.typemap.items():
+        new_var = ir.Var(scope, mk_unique_var(v), loc)
+        var_dict[v] = new_var
+        typemap[new_var.name] = typ  # add new var type for overall function
+    ir_utils.replace_vars(stencil_blocks, var_dict)
+
+    if config.DEBUG_ARRAY_OPT == 1:
+        print("After replace_vars")
+        ir_utils.dump_blocks(stencil_blocks)
+
+    # add call types to overall function
+    for call, call_typ in tp.calltypes.items():
+        calltypes[call] = call_typ
+
+    arg_to_arr_dict = {}
+    # TODO: handle closure vars
```

Comment:
This is not applicable anymore since the stencil function is not compiled here.

Earlier, I was considering closure variables coming from the outside function in the stencil kernel definition. However, supporting closures might not make sense or be possible with all the functionality/options we have added.